### ISSUE TITLE
Assert in Number.FormatDouble

### ISSUE THREAD

**Issue by stephentoub** [dotnet/coreclr#25081]
I ran the System.Runtime.Tests (`dotnet msbuild /t:rebuildandtest /p:XunitOptions="-maxthreads 512"`) using a locally built checked coreclr from master, and hit the following assert:
```
    Discovering: System.Runtime.Tests (method display = ClassAndMethod, method display options = None)
    Discovered:  System.Runtime.Tests (found 5336 of 5369 test cases)
    Starting:    System.Runtime.Tests (parallel test collections = on, max threads = 512)
  Process terminated. Assertion failed.
     at System.Number.FormatDouble(ValueStringBuilder& sb, Double value, ReadOnlySpan`1 format, NumberFormatInfo info)
     at System.Number.FormatDouble(Double value, String format, NumberFormatInfo info)
     at System.Double.ToString(String format, IFormatProvider provider)
     at System.Tests.RealFormatterTests.InvariantToStringDouble(Double d, String format) in d:\repos\corefx\src\System.Runtime\tests\System\RealFormatterTests.netcoreapp.cs:line 15
     at System.Tests.RealFormatterTestsBase.TestFormatterDouble_Standard(Double value, String format, String expectedResult) in d:\repos\corefx\src\Common\tests\System\RealFormatterTestsBase.netcoreapp.cs:line 301
     at System.Tests.RealFormatterTestsBase.TestFormatterDouble_N20(Double value, String expectedResult) in d:\repos\corefx\src\Common\tests\System\RealFormatterTestsBase.netcoreapp.cs:line 210
     at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
     at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
     at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
```
cc: @tannergooding

**Comment by stephentoub** [comment]
Ran it again, this time with PDBs:
```
    Discovering: System.Runtime.Tests (method display = ClassAndMethod, method display options = None)
    Discovered:  System.Runtime.Tests (found 5336 of 5369 test cases)
    Starting:    System.Runtime.Tests (parallel test collections = on, max threads = 512)
  Process terminated. Assertion failed.
     at System.Number.FormatDouble(ValueStringBuilder& sb, Double value, ReadOnlySpan`1 format, NumberFormatInfo info) in d:\repos\coreclr\src\System.Private.CoreLib\shared\System\Number.Formatting.cs:line 564
     at System.Number.FormatDouble(Double value, String format, NumberFormatInfo info) in d:\repos\coreclr\src\System.Private.CoreLib\shared\System\Number.Formatting.cs:line 382
     at System.Double.ToString(String format, IFormatProvider provider) in d:\repos\coreclr\src\System.Private.CoreLib\shared\System\Double.cs:line 257
     at System.Tests.RealFormatterTests.InvariantToStringDouble(Double d, String format) in d:\repos\corefx\src\System.Runtime\tests\System\RealFormatterTests.netcoreapp.cs:line 15
     at System.Tests.RealFormatterTestsBase.TestFormatterDouble_Standard(Double value, String format, String expectedResult) in d:\repos\corefx\src\Common\tests\System\RealFormatterTestsBase.netcoreapp.cs:line 301
     at System.Tests.RealFormatterTestsBase.TestFormatterDouble_N20(Double value, String expectedResult) in d:\repos\corefx\src\Common\tests\System\RealFormatterTestsBase.netcoreapp.cs:line 210
     at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
```

**Comment by tannergooding** [comment]
Are you using the latest sources? The line number appears to be off by 2 (https://source.dot.net/#System.Private.CoreLib/shared/System/Number.Formatting.cs,562).

I'll see if I can reproduce locally.

**Comment by stephentoub** [comment]
> Are you using the latest sources?

Yes, latest in both coreclr and corefx.  Pulled from master from both this morning.

**Comment by jkotas** [comment]
Any chance to get a dump of the crash? https://github.com/dotnet/coreclr/issues/25080#issuecomment-500937525

**Comment by stephentoub** [comment]
Shared a dump offline.

**Comment by stephentoub** [comment]
Here are the values causing the assert to fail:
```
  precision: 20
  DoublePrecision: 17
  BitConverter.DoubleToInt64Bits(value): 1
  BitConverter.DoubleToInt64Bits(NumberToDouble(ref number)): 0
```

**Comment by tannergooding** [comment]
Yep. I was able to repro locally and determine the root cause.

This was introduced as part of https://github.com/dotnet/coreclr/commit/447670fb7cfbf0812faa7cbf4792195d56844de5. Namely, we should be preserving the first significant digit in the scenario where it would not cause the next digit to round. This avoids us unnecessarily treating a value as zero when it is actually epsilon.

**Comment by stephentoub** [comment]
Thanks, @tannergooding.  I'm curious why this assert doesn't fire in coreclr's CI when running corefx tests, since I thought they were run using a checked build... is the test triggering this just too new to be in coreclr's copy?

**Comment by tannergooding** [comment]
Yes, that is likely the case. I believe CoreCLR is still running against a stale copy of CoreFX, but @jkotas  might know if that has changed.
-- Maybe it would be good for us to have an outer loop job in CoreFX that runs against a checked build of CoreCLR?

**Comment by jkotas** [comment]
CoreCLR is running against current CoreFX binaries and tests. The CI has only one "Windows_NT x64_release" leg enabled right now. @BruceForstall is working to add more add in #24962

**Comment by BruceForstall** [comment]
Excellent! fwiw, it was a huge pain to figure out how to disable these failures in #24962

**Comment by stephentoub** [comment]
> it was a huge pain to figure out how to disable these failures

Were there other failures #24962 caught that we should know about? :wink:

**Comment by BruceForstall** [comment]
@stephentoub I just merged #24962. And opened #25098 to track all the exclusions I added.

Summary:
1. System.Net has lots of problems with timeouts or other infrastructure-looking failures.
2. There are definitely a few real failures that need to be investigated.
3. I opened 2 other issues for VM/JIT asserts.
